### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by FelixReuss**

### CODE REVIEW

File: src/colorella/colormap.py [hunk 1]
```diff
@@ -41,85 +42,49 @@
...
-class ColorMap:  # TODO: class names start always with an upper case, e.g. ColorMap
+class ColorMap:
     """create a colormap object compatible with matplotlib
-    TODO: implement alpha channel support in load and save method. Should colormap.objects be in place or a new object
         """
 
-    def __init__(self, arg, name='default', cm_dirpath='./colormaps'):
-        # TODO: blank between parameter and :
-        # TODO: str or list or dict, optional
+    def __init__(self, arg, cm_dirpath='./colormaps', name=None):
```
Comment (sebhahn) [inline src/colorella/colormap.py]: the path is not independent for the user OS

File: src/colorella/colormap.py [hunk 2]
```diff
@@ -206,21 +169,48 @@ def save_as_cpt(self, outname=None, **kwargs):
...
-            os.makedirs(outname)
-        np.savetxt(outname, arr, fmt=fmt,
+        np.savetxt(outname, col_arr, fmt=fmt,
                    header="# COLOR_MODEL = RGB",
-                   footer=ex, comments="")
+                   footer=footer, comments="")
+
+    def save_as_ct(self, outname=None):
+        """
+        Saves a acolormap.object as a gdal .ct file
```
Comment (sebhahn) [inline src/colorella/colormap.py]: typo

File: src/colorella/colormap.py [hunk 3]
```diff
@@ -206,21 +169,48 @@ def save_as_cpt(self, outname=None, **kwargs):
...
+            outname = os.path.join(self.dirpath, self._mpl_cm.name+'.ct')
+        elif os.path.isdir(outname):
+            outname = os.path.join(outname, self._mpl_cm.name+'.ct')
+        elif '.ct' not in outname:
+            outname = outname+'.ct'
+
+        arr = np.zeros((255, 3), dtype=int)
+
+        if isinstance(self._mpl_cm, col.ListedColormap):
+            for i in range(len(self._mpl_cm.colors)):
```
Comment (sebhahn) [inline src/colorella/colormap.py]: can this be done without a loop? `self._mpl_cm.colors` is it an array or list?

File: src/colorella/colormap.py [hunk 4]
```diff
@@ -348,29 +335,171 @@ def reverse(self, inplace=True):
...
+            name, cptdict = cptfile2dict(filepath)
+            mpl_cm = col.LinearSegmentedColormap(name=name, segmentdata=cptdict)
+        elif '.ct' == extension:
+            name, gdallist = ctfile2dict(filepath)
+            mpl_cm = col.LinearSegmentedColormap.from_list(name=name, colors=gdallist)
+        elif '.json' == extension:
+            name, jsonlist = json2list(filepath)
+            mpl_cm = col.ListedColormap(jsonlist, name=name)
+        else:
+            raise Exception('')
```
Comment (sebhahn) [inline src/colorella/colormap.py]: Exception is too general and please add a message

**Review by sebhahn — CHANGES_REQUESTED** [review]
Please run a pep8 check of the code (using e.g. autopep8)

**Comment by sebhahn** [comment]
The tests are also not working

**Comment by sebhahn** [comment]
I also talked to @cnavacch, it would be good to make the dependency of `gdal` optional, i.e. gdal functions can only be used if the package has been installed before, but is not installed by default

**Comment by cnavacch** [comment]
@sebhahn : Thanks for the reviews, but we should have informed you that this is only a preliminary PR. Felix and I did not had the chance to speak to each other to apply the changes we have discussed the last time. Felix will let you know as soon as the changes are deployed.

**Comment by sebhahn** [comment]
Ok, but the tests should be working for this PR.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+[![PyPi Package](https://badge.fury.io/py/colorella.svg)](http://badge.fury.io/py/colorella)
+[![RTD](https://readthedocs.org/projects/colorella/badge/?version=latest)](http://colorella.readthedocs.org/)
+[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
+
+## Description
+*Colorella* stands for **Col**or **or**ganizing and **e**asy to **l**earn **la**boratory
+and provides a colormap class that allows the usage of colormaps from different sources and functions for their efficient application
+
+## Installation
+
```
Comment (cnavacch) [inline README.md]: Needs further descriptions, please have a look at other packages.

File: src/colorella/colormap.py [hunk 6]
```diff
@@ -18,209 +18,226 @@
...
 
-from src.colorella.utils import cptfile2dict, gdal2dict, json2list
-
+gdal_warning = 'No GDAL Installation found. Without gdal the following functions are not available: from_gdal, to_gdal'
+try:
+    from osgeo import gdal
+    GDAL_INSTALLED = True
+except:
+    GDAL_INSTALLED = False
+    raise warnings.warn(gdal_warning)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: I would only raise the warning when the user tries to execute one of the functions

File: src/colorella/colormap.py [hunk 7]
```diff
@@ -18,209 +18,226 @@
...
-            warnings.warn("Colormap is already a Segmented Colormap. Listed Colormap required")
-            return self._mpl_cm
-
-        else:
-            mpl_cm = col.LinearSegmentedColormap.from_list(name, self._mpl_cm.colors)
+        name, colors, gradient = json2list(filepath)
+        if gradient == False:
+            return cls.from_list(colors, name=name, gradient=gradient)
+        if gradient == True:
+            return cls.from_dict(colors, name=name)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: You do not need to check a truth value, use:
```
        if gradient:
             return cls.from_dict(colors, name=name)
        else:
             return cls.from_list(colors, name=name, gradient=gradient)
```

File: src/colorella/colormap.py [hunk 8]
```diff
@@ -247,14 +263,52 @@ def convert2greyscale(self, weights=1, inplace=True):
             warnings.warn('Argument weight only supports values between 1 and 3')
 
         colors[:, :3] = luminance[:, np.newaxis]
-        mpl_cm = col.LinearSegmentedColormap.from_list(self._name + "_grey", colors, self._mpl_cm.N)
+        mpl_cm = col.LinearSegmentedColormap.from_list(self._mpl_cm.name +'_grey', colors, self._mpl_cm.N)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: You can call `self.name` instead of `self._mpl_cm.name`

File: src/colorella/colormap.py [hunk 9]
```diff
@@ -247,14 +263,52 @@ def convert2greyscale(self, weights=1, inplace=True):
...
+                keys.append(key)
+                channel = self._mpl_cm._segmentdata[key]
+                data = []
+
+                for c in channel:
+                    data.append((1 - c[0], c[2], c[1]))
+                reverse.append(sorted(data))
+
+            revdict = dict(zip(keys, reverse))
+            mpl_cm = mpl.colors.LinearSegmentedColormap(segmentdata=revdict, name=self._mpl_cm.name + '_reversed')
```
Comment (cnavacch) [inline src/colorella/colormap.py]: You can call `self.name` instead of `self._mpl_cm.name`

File: src/colorella/colormap.py [hunk 10]
```diff
@@ -309,68 +361,174 @@ def to_list(self):
...
         """
-        colors = self._mpl_cm(np.arange(self._mpl_cm.N))
-        plt.imshow([colors], extent=[0, 10, 0, 1])
-        plt.axis('off')
-        plt.show()
+        if isinstance(self._mpl_cm, col.LinearSegmentedColormap):
+            warnings.warn("Colormap is already a Segmented Colormap. Listed Colormap required")
+            return self
+        else:
+            mpl_cm = col.LinearSegmentedColormap.from_list(name=self._mpl_cm.name+'_gradient', colors=self._mpl_cm.colors)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: You can call `self.name` instead of `self._mpl_cm.name`

File: src/colorella/colormap.py [hunk 11]
```diff
@@ -18,209 +18,226 @@
...
 import matplotlib.cm as cm
 import matplotlib.colors as col
-from osgeo import gdal
+import colorcet as cc
 import warnings
+from colorella.conversions import cptfile2dict, ctfile2dict, json2list
 
-from src.colorella.utils import cptfile2dict, gdal2dict, json2list
-
+gdal_warning = 'No GDAL Installation found. Without gdal the following functions are not available: from_gdal, to_gdal'
```
Comment (cnavacch) [inline src/colorella/colormap.py]: Fine to define it here, but write it in capital letters. This makes it clear, that it is a global variable

File: src/colorella/colormap.py [hunk 12]
```diff
@@ -309,68 +361,174 @@ def to_list(self):
...
+                    gdal_ct.SetColorEntry(int(i / accelerate), tuple(np.rint(np.multiply(mpl_ct[i], 256)).astype(np.byte)) + (0,))
+                for i in range(256 // accelerate, 256):
+                    gdal_ct.SetColorEntry(i, tuple((255, 255, 255)) + (0,))
+            elif isinstance(self._mpl_cm, col.LinearSegmentedColormap):
+                mpl_ct = (self._mpl_cm(np.linspace(0., 1., 255))[:, :3] * 255).astype(int)
+                for i in range(0, mpl_ct.shape[0], accelerate):
+                    gdal_ct.SetColorEntry(int(i / accelerate),
+                                          tuple(np.rint(np.multiply(mpl_ct[i], 256)).astype(np.byte)) + (0,))
+                for i in range(256 // accelerate, 256):
+                    gdal_ct.SetColorEntry(i, tuple((255, 255, 255)) + (0,))
```
Comment (cnavacch) [inline src/colorella/colormap.py]: Maybe you move this part to the conversion.py file?

**Review by cnavacch — CHANGES_REQUESTED** [review]
Please fix my comments before merging. I have three additional comments, which I was not able to mark in the PR:

1. What is "src/colorella/colormaps/etopo1.cpt " doing under colormaps? There should only be ct files.
2. Some functions still do not have a proper NumPy Style documentation, please fix this.
3. Sometimes you use a `try` and `except` block to read a file. I would prefer to use `os.path.exists()`.
4. Have you setup ReadTheDocs already? I would also do this before merging. 
5. Sometimes you have a camel-case variable naming. Please use lower letters with underscores to follow the PEP-8 convention.

File: tests/test_colormap.py [hunk 13]
```diff
@@ -23,10 +23,17 @@
...
 from colorella.colormap import ColorMap
 import matplotlib.colors as col
+import warnings
+gdal_warning = 'No GDAL Installation found. Without gdal the following functions are not available: from_gdal, to_gdal'
+try:
+    from osgeo import gdal
+    GDAL_INSTALLED = True
+except:
+    GDAL_INSTALLED = False
+    raise warnings.warn(gdal_warning)
```
Comment (cnavacch) [inline tests/test_colormap.py]: This should only be raised when a gdal function is explicitly called. It is enough when you define it in the colormap class and catch the warning in the test class.

**Review by cnavacch — CHANGES_REQUESTED** [review]

File: src/colorella/colormap.py [hunk 14]
```diff
@@ -18,209 +18,303 @@
...
+        ColorMap object
         """
-        Returns the xth color of the colormap
+
+        mpl_cm = None
+        extension = os.path.splitext(filepath)[1]
+        if '.cpt' == extension:
+            filename, cpt_list, cpt_dict = cptfile2dict(filepath)
+            name = name if name is not None else filename
+            if gradient == False:
```
Comment (cnavacch) [inline src/colorella/colormap.py]: Again, you evaluate a boolean to `== true` or `== false`. Use if `gradient` or `if not gradient.`

File: README.md [hunk 15]
```diff
@@ -0,0 +1,1331 @@
...
+
+
+
+
+<!DOCTYPE html>
```
Comment (cnavacch) [inline README.md]: This should not be here as an html file. if you want to integrate the graphic use png or jpeg, not plain SVG.

File: docs/index_old.rst [hunk 16]
```diff
@@ -0,0 +1,80 @@
+=========
```
Comment (cnavacch) [inline docs/index_old.rst]: please use one index file, not two.

File: src/colorella/conversions.py [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+                    colors = [tuple(i) for i in rgb]
+
+            else:
+                colors = [cmap_dict['RGBPoints'][x:x + 3] for x in range(0, len(cmap_dict['RGBPoints']), 4)]
+            if cmap_dict.get('RGBPoints', None) is None:
+                return None
+            return name, colors, gradient
+    else:
+        raise ImportError("file ", filepath, "not found")
+        return None
```
Comment (cnavacch) [inline src/colorella/conversions.py]: not really necessary when error is thrown before

File: src/colorella/conversions.py [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+        color_model = "RGB"
+        for l in lines:
+            ls = l.split()
+            if l.strip():
+                if l[0] == "#":
+                    if ls[-1] == "HSV":
+                        color_model = "HSV"
+                        continue
+                    else:
+                        continue
```
Comment (cnavacch) [inline src/colorella/conversions.py]: You could shorten this by saying:
```                    
 if ls[-1] == "HSV":
     color_model = "HSV"
 continue
```

File: src/colorella/colormap.py [hunk 19]
```diff
@@ -18,209 +18,303 @@
...
+        elif '.json' == extension:
+            filename, colors, gradient = json2list(filepath)
+            name = name if name is not None else filename
+            if gradient == False:
+                mpl_cm = col.ListedColormap(colors,  name=name)
+            if gradient == True:
+                mpl_cm = col.LinearSegmentedColormap(name=name, segmentdata=colors)
+
+        else:
+            raise ValueError('File extensions is not recognized, supported file extensions are: .cpt, .ct, .json')
```
Comment (cnavacch) [inline src/colorella/colormap.py]: You could shorten this by putting the inner gradient if clause outside and after the outer if clause.

File: src/colorella/colormap.py [hunk 20]
```diff
@@ -18,209 +18,303 @@
...
+        ColorMap object (LinearSegmented Colormap object)
         """
-        Returns a string containing the RGB values of the colormap
-        - for Segmented Colormap: Start and end color
-        - for a Listed Colormap: all colors
+        name, cpt_list, cpt_dict = cptfile2dict(filepath)
+        if gradient == False:
+            return cls.from_list(cpt_list, name=name, gradient=False)
+        if gradient == True:
+            return cls.from_dict(cpt_dict, name=name)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: again, do the boolean check without `==`

**Review by cnavacch — CHANGES_REQUESTED** [review]

File: src/colorella/colormap.py [hunk 21]
```diff
@@ -50,7 +50,7 @@
     GDAL_INSTALLED = True
 except:
     GDAL_INSTALLED = False
-    raise warnings.warn(gdal_warning)
+    warnings.warn(gdal_warning)
```
Comment (cnavacch) [inline src/colorella/colormap.py]: I would only print a warning in the GDAL functions.

File: src/colorella/conversions.py [hunk 22]
```diff
@@ -177,7 +175,6 @@ def json2list(filepath):
             return name, colors, gradient
     else:
         raise ImportError("file ", filepath, "not found")
-        return None
```
Comment (cnavacch) [inline src/colorella/conversions.py]: Could you please remove this also for the other file processes where an ImportError is raised?

**Review by cnavacch — CHANGES_REQUESTED** [review]

**Review by cnavacch — APPROVED** [review]

**Comment by cnavacch** [comment]
@sebhahn Could you please take another look at the code and give some final feedback? Felix can then finally complete the PR after he resolved your comments :-)

File: docs/conf.py [hunk 23]
```diff
@@ -1,273 +1,60 @@
...
-output_dir = os.path.join(__location__, "api")
-module_dir = os.path.join(__location__, "../src/colorella")
-try:
-    shutil.rmtree(output_dir)
-except FileNotFoundError:
-    pass
+#
+import os
+import sys
+sys.path.insert(0, os.path.abspath('H:\GIT\colorella'))
```
Comment (sebhahn) [inline docs/conf.py]: Why is this path in here and hardcoded?

File: docs/conf.py [hunk 24]
```diff
@@ -1,273 +1,60 @@
...
-    cmd_line = cmd_line_template.format(
-        outputdir=output_dir, moduledir=module_dir)
+# -- Project information -----------------------------------------------------
 
-    args = cmd_line.split(" ")
-    if parse_version(sphinx.__version__) >= parse_version('1.7'):
-        args = args[1:]
+project = 'colorella'
+copyright = '2020, Felix Reuss'
+author = 'Felix Reuss'
```
Comment (sebhahn) [inline docs/conf.py]: We usually use 'TU Wien' as author and copyright

File: src/colorella/conversions.py [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+        for i in range(len(x)):
+            red.append([x_norm[i], r[i], r[i]])
+            green.append([x_norm[i], g[i], g[i]])
+            blue.append([x_norm[i], b[i], b[i]])
+
+        color_dict = {"red": red, "green": green, "blue": blue}
+
+        return name, col_list, color_dict
+    else:
+        raise ImportError("file ", filepath, "not found")
```
Comment (sebhahn) [inline src/colorella/conversions.py]: I would recommend

```
if not os.path.exists(filepath):
    raise ImportError("file ", filepath, "not found")

# no else
file = open(filepath)
name = os.path.splitext(os.path.basename(filepath))[0]
lines = file.readlines()
file.close()
...
return ...
```

File: src/colorella/conversions.py [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+        blue = np.array(blue, dtype=np.float64)
+
+        red = red/255
+        green = green/255
+        blue = blue/255
+
+        col_list = [(red[i], green[i], blue[i]) for i in range(len(red))]
+        return name, col_list
+
+    else:
```
Comment (sebhahn) [inline src/colorella/conversions.py]: see comment above

File: src/colorella/conversions.py [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+                else:
+                    rgb = cmap_dict['RGBPoints']
+                    colors = [tuple(i) for i in rgb]
+
+            else:
+                colors = [cmap_dict['RGBPoints'][x:x + 3] for x in range(0, len(cmap_dict['RGBPoints']), 4)]
+            if cmap_dict.get('RGBPoints', None) is None:
+                return None
+            return name, colors, gradient
+    else:
```
Comment (sebhahn) [inline src/colorella/conversions.py]: see comment above

**Review by sebhahn — CHANGES_REQUESTED** [review]
Some minor comments

**Pull request merged**